### PR TITLE
Add `ArrayAt` type

### LINKED ISSUE #369: Add `ArrayAt` type for stricter `Array.prototype.at()` with tuples
See https://github.com/sindresorhus/ts-extras/issues/39 and https://github.com/microsoft/TypeScript/issues/47660.

```ts
const tuple = ['abc', 123, true] as const;

type First = ArrayAt<typeof tuple, 0>; // 'abc'
type Last = ArrayAt<typeof tuple, -1>; // true
type SecondToLast = ArrayAt<typeof tuple, -2>; // 123
type ThirdToLast = ArrayAt<typeof tuple, -3>; // 'abc'
type OutOfBounds = ArrayAt<typeof tuple, 999>; // should not compile since 999 does not extend 0 | 1 | 2 | -1 | -2 | -3
```

```ts
const array = ['abc', 123, true];

type First = ArrayAt<typeof tuple, 0>; // string | number | true
type Last = ArrayAt<typeof tuple, -1>; // string | number | true
type OutOfBounds = ArrayAt<typeof tuple, 999>; // string | number | true
```

### PR DESCRIPTION
**Pull request by giao66**
closed #369

### CODE REVIEW

File: test-d/array-at.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+expectType<ArrayAt<LeadingSpreadArray, 0>>({} as string | number);
+expectType<ArrayAt<LeadingSpreadArray, 1>>({} as string | boolean);
+expectType<ArrayAt<LeadingSpreadArray, 2>>({} as string | undefined);
+expectType<ArrayAt<LeadingSpreadArray, 3>>({} as string | undefined);
+
+expectType<ArrayAt<TrailingSpreadArray, 0>>({} as number);
+expectType<ArrayAt<TrailingSpreadArray, 1>>({} as boolean);
+expectType<ArrayAt<TrailingSpreadArray, 2>>({} as string | undefined);
+expectType<ArrayAt<TrailingSpreadArray, 3>>({} as string | undefined);
```
Comment (som-sm) [inline test-d/array-at.ts:25]: Please add more tests, for example tests with negative indicies, optional elements with spread, union of arrays, union of indicies etc, look at other tests in the repo for inspiration.

From quick testing, the current implementation fails in the following cases:
```ts
type Current = ArrayAt<[...string[], number], -1>; // string | number

type Expected = ArrayAt<[...string[], number], -1>; // number
```

```ts
type Current = ArrayAt<[boolean, number?, ...string[]], 1>; // string | number | undefined
type Expected = ArrayAt<[boolean, number?, ...string[]], 1>; // number | undefined
```

Comment (giao66) [reply test-d/array-at.ts:39]: ok! handsome guy

File: test-d/array-at.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+expectType<ArrayAt<TrailingSpreadArray, -2>>({} as boolean | string | number);
+expectType<ArrayAt<TrailingSpreadArray, -3>>({} as boolean | string | number | undefined);
+
+expectType<ArrayAt<TrailingSpreadArray2, 0>>({} as object);
+expectType<ArrayAt<TrailingSpreadArray2, 1>>({} as number);
+expectType<ArrayAt<TrailingSpreadArray2, 2>>({} as string | boolean | undefined);
+expectType<ArrayAt<TrailingSpreadArray2, 3>>({} as string | boolean | undefined);
+expectType<ArrayAt<TrailingSpreadArray2, -1>>({} as string | number | boolean | undefined);
+expectType<ArrayAt<TrailingSpreadArray2, -2>>({} as string | number | boolean | object | undefined);
+expectType<ArrayAt<TrailingSpreadArray2, -3>>({} as string | number | boolean | object | undefined);
```
Comment (sindresorhus) [inline test-d/array-at.ts:50]: More tests:

```ts
type StaticArray3 = [string, number?, boolean?];
expectType<ArrayAt<StaticArray3, -1>>({} as boolean | undefined);
expectType<ArrayAt<StaticArray3, -2>>({} as number | undefined);

type EmptyArray = [];
expectType<ArrayAt<EmptyArray, 0>>(undefined);
expectType<ArrayAt<EmptyArray, -1>>(undefined);

type SingleElement = [string];
expectType<ArrayAt<SingleElement, 0>>({} as string);
expectType<ArrayAt<SingleElement, -1>>({} as string);

type UnionArray = [string, number] | [boolean, string?];
expectType<ArrayAt<UnionArray, 0>>({} as string | boolean);
expectType<ArrayAt<UnionArray, 1>>({} as number | string | undefined);
expectType<ArrayAt<UnionArray, -1>>({} as number | string | boolean | undefined);
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
There are still unhandled feedback.

Resolve the comments when you are sure they are resolved.

**Comment by WeiWen1999** [comment]
> There are still unhandled feedback.
> 
> Resolve the comments when you are sure they are resolved.

HELLO! I DONE

File: source/array-at.d.ts [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+type C = ArrayAt<A, -1>;
+//=> boolean
+*/
+export type ArrayAt<T extends UnknownArray, N extends number> =
+number extends T['length']
+	? NonFixedLengthArrayAt<T, N>
+	: FixedLengthArrayAt<T, N>;
+
+type FixedLengthArrayAt2<T extends UnknownArray, N extends number> =
```
Comment (sindresorhus) [inline source/array-at.d.ts]: This is unused. There are also some other unused utilities.

File: source/internal/type.d.ts [hunk 4]
```diff
@@ -106,3 +106,8 @@ Indicates the value of `exactOptionalPropertyTypes` compiler option.
 export type IsExactOptionalPropertyTypesEnabled = [(string | undefined)?] extends [string?]
 	? false
 	: true;
+
+/**
+Return the value of exactOptionalPropertyTypes option in tsconfig
+*/
+export type ExactOptionalPropertyTypesEnable = [(string | undefined)?] extends [string?] ? false : true;
```
Comment (sindresorhus) [inline source/internal/type.d.ts]: This is already defined above.

**Comment by sindresorhus** [comment]
You force-pushed over my changes.

**Comment by WeiWen1999** [comment]
> You force-pushed over my changes.

So I should sync the main branch first before pushing, right?

**Comment by WeiWen1999** [comment]
> You force-pushed over my changes.

Sorry，This is my first time trying to contribute to an open-source project.

**Comment by giao66** [comment]
Can this version be restored?
p

**Comment by WeiWen1999** [comment]
> You force-pushed over my changes.

Can this version be restored?
p

**Comment by sindresorhus** [comment]
It's fine. I will do the changes again after merging.

**Comment by sindresorhus** [comment]
AI review:

---

1. **Stray debug/types in the public file**

   * `source/array-at.d.ts` contains scratch types like `type TrailingSpreadArray2`, `type MiddleSpreadArray`, `type A/B/C`, etc. These should never ship in a public `.d.ts`. Remove them. ([[GitHub](https://github.com/sindresorhus/type-fest/pull/1105/files)][1])

2. **Dead import(s)**

   * `GreaterThan` is imported but not used (I only see `GreaterThanOrEqual`/`LessThan`/`LessThanOrEqual` being referenced). Drop it. Keep imports minimal to reduce checker work. ([[GitHub](https://github.com/sindresorhus/type-fest/pull/1105/files)][1])

3. **Unknown index handling (generic `N`)**

   * Top level signature allows `N extends number`. If `N` is not a literal, comparisons like `IsNegative<N>`/`LessThan<N, …>` become indeterminate and you get brittle distributive behavior or over-narrowing.
   * Add a fast path: for unknown `N`, return a safe union.

     * For fixed-length tuples: `number extends N ? (IsExactOptionalPropertyTypesEnabled extends true ? Required<T>[number] : T[number]) | undefined : …`
     * For non-fixed arrays: `number extends N ? T[number] : …`
   * This mirrors runtime: unknown index => “some element (maybe) or `undefined`”. It also prevents the type from collapsing to `never` in some branches.

# Potential logic gaps to double-check

* **Fixed tuple, out-of-bounds negative**: The `Sum<T['length'], N>` branch correctly yields `undefined` when the computed index is still negative. Keep that, but cover it with tsd (e.g., `ArrayAt<[1, 2, 3], -4> -> undefined`). ([[GitHub](https://github.com/sindresorhus/type-fest/pull/1105/files)][1])
* **Trailing spread, negative indices**: The unioning via the “slice then `…infer Last`” approach looks right (e.g., `[A, B, ...C[]], -1 => B | C | undefined` depending on spread arity), but you must assert it with tests—especially for `-2`, `-3`, etc. where unions widen quickly. ([[GitHub](https://github.com/sindresorhus/type-fest/pull/1105/files)][1])
* **Leading spread, positive indices**: The positive branch unions the possible holders for that position: `ArraySlice<StaticPart, 0, N+1>[number] | VariablePart[number] | (N≥len ? undefined : never)`. This is conceptually correct; again, codify with tests. ([[GitHub](https://github.com/sindresorhus/type-fest/pull/1105/files)][1])
* **Middle spread recursion**: The recursion to `NonFixedLengthArrayAt<[...VariablePart, ...TrailingStaticPart], …>` (for positive) and the symmetric negative handling looking into `TrailingStaticPart` is good, but performance-sensitive. Make sure tsd stays fast on large tuples (see perf note below). ([[GitHub](https://github.com/sindresorhus/type-fest/pull/1105/files)][1])

# Tests you’re missing (add to `test-d/array-at.ts`)

Cover each of these explicitly with `tsd` assertions:

**Fixed-length tuples**

* `ArrayAt<[], 0> -> undefined`
* `ArrayAt<[A], 0> -> A`
* `ArrayAt<[A], -1> -> A`
* `ArrayAt<[A, B, C], -1> -> C`
* `ArrayAt<[A, B, C], -4> -> undefined`
* Optional elements with and without `exactOptionalPropertyTypes`:

  * `ArrayAt<[A?, B], 0>` under both modes
  * `ArrayAt<[A?, B?], -1>` under both modes

**Non-fixed arrays**

* `ArrayAt<string[], 0> -> string`
* `ArrayAt<string[], -1> -> string | undefined` (document the choice you want; if you keep it as just `string`, state why)
* `ArrayAt<readonly number[], 1> -> number` (readonly preservation)

**Spread patterns**

* Leading: `ArrayAt<[...string[], A, B], 0 | 1 | 2 | -1>`
* Middle: `ArrayAt<[A, ...string[], B], 0 | 1 | 2 | -1 | -2>`
* Trailing: `ArrayAt<[A, B, ...string[]], -1 | -2 | -3>`
* Mixed optional statics like `[A?, ...string[], B?]` near boundaries

**Unknown index**

* `ArrayAt<[A, B, C], number>` – should be `A | B | C | undefined` (or at least `A | B | C`; pick and document)
* `ArrayAt<string[], number>` – `string`

# Cleanups / polish

* Remove the stray scratch types from `source/array-at.d.ts`. ([[GitHub](https://github.com/sindresorhus/type-fest/pull/1105/files)][1])
* Remove unused `GreaterThan` import. ([[GitHub](https://github.com/sindresorhus/type-fest/pull/1105/files)][1])
* Consider extracting tiny helpers to reduce repetition (`NormalizeIndex<T,N>`, etc.) if it doesn’t regress checker perf.

# Docs

* README entry is fine, but add one more example showing a negative index on a tuple and on a trailing-spread case, plus a note that for non-fixed arrays the result is a union (and may or may not include `undefined`, per your chosen semantics). ([[GitHub](https://github.com/sindresorhus/type-fest/pull/1105/files)][1])

# Perf

* This kind of type explodes unions quickly. Run the usual `tsc --extendedDiagnostics` locally on the repo’s test suite once you add the broader cases. If anything spikes, prefer fewer nested `extends infer` steps in the hot branches.

---

## TL;DR required changes before merge

* Delete the debug types from `source/array-at.d.ts`. **Blocker.** ([[GitHub](https://github.com/sindresorhus/type-fest/pull/1105/files)][1])
* Drop unused `GreaterThan` import. **Blocker.** ([[GitHub](https://github.com/sindresorhus/type-fest/pull/1105/files)][1])
* Add an explicit **unknown-index guard** for `N extends number` (see above). **Strongly recommended.**
* Add the **missing tsd cases** outlined above, especially around `exactOptionalPropertyTypes`. **Strongly recommended.** ([[TypeScript](https://www.typescriptlang.org/tsconfig/exactOptionalPropertyTypes.html?utm_source=chatgpt.com)][2])

[1]: https://github.com/sindresorhus/type-fest/pull/1105/files "Add `ArrayAt` type by giao66 · Pull Request #1105 · sindresorhus/type-fest · GitHub"
[2]: https://www.typescriptlang.org/tsconfig/exactOptionalPropertyTypes.html?utm_source=chatgpt.com "TSConfig Option: exactOptionalPropertyTypes"
[3]: https://stackoverflow.com/questions/50490773/why-doesnt-typescript-undefined-type-behave-same-as-optional?utm_source=chatgpt.com "Why doesn't typescript undefined type behave same as ..."

**Comment by sindresorhus** [comment]
After doing the above, I recommend asking ChatGPT to review your pull request and ask it to add more tests if needed.

**Comment by WeiWen1999** [comment]
> After doing the above, I recommend asking ChatGPT to review your pull request and ask it to add more tests if needed.

ok,wizard.Thanks a lot!

**Comment by sindresorhus** [comment]
* **Wrong `Every` docs (copy-paste).** The block under `export type Every…` documents `CollapseRestElement` with unrelated examples. Replace with correct description and examples for `Every<TArray, Type>`. ([GitHub][1])

* **Misuse of `IsEqual` (no type args).** In `StaticPartOfArray`, you wrote `IsEqual extends true` without parameters, which is invalid. Use an actual comparison, e.g., compare the head optional element:

  ```ts
  type HeadOptional =
  	OptionalPartOfStaticArray<T> extends [infer O] ? O : never;

  type HasOptionalHead = IsEqual<HeadOptional, undefined>;
  // …
  : HasOptionalHead extends true ? Result : [...Result, Required<T>[0]?]
  ```

(Adjust to your existing generics for `T`, `Result`.) ([GitHub][1])

* **Duplicate index computation (leading-spread, negative index).** In `NonFixedLengthArrayAt`’s leading-spread negative branch you compute `Sum<StaticPart['length'], N> extends infer Index` twice. Compute once and reuse. ([GitHub][1])

* **Doc wording mismatch.** `LeadingStaticPartOfMiddleSpreadArray` comment says “leading **variable**, non-fixed-length portion,” but the name is “StaticPart.” Change “variable” → “static” to avoid confusion. ([GitHub][1])

* **Missing internal re-exports.** `source/array-at.d.ts` imports many helpers from `./internal/index.d.ts`, but this PR doesn’t add those exports to that aggregator. Add them or import directly from `./internal/array.d.ts` to prevent broken imports. ([GitHub][1])

* **Examples in `ArrayAt` JSDoc are wrong/incomplete.** They show `type B = ArrayAt;` without type parameters. Use concrete, compilable examples (both positive and negative indices):

  ````ts
  type B = ArrayAt<[string, number, boolean], 0>;  // string
  type C = ArrayAt<[string, number, boolean], -1>; // boolean
  type D = ArrayAt<string[], -1>;                  // string | undefined
  ``` :contentReference[oaicite:5]{index=5}

  ````

* **Over-defensive conditional in fixed-tuple branch.** `Index extends unknown ? … : never` is redundant (always true). Drop the extra layer for readability.

* **Test coverage gaps.** Add:

  * `readonly` tuples.
  * Union index literals (`0 | 2`, `-1 | -3`).
  * Union of array kinds (`string[] | number[]`).
  * Negative out-of-bounds on empty tuple (`ArrayAt<[], -1>`).
    Current tests cover many shapes but miss these. ([GitHub][1])

* **Minor housekeeping.** Keep the `index.d.ts` export grouped with other array utilities (`ArraySlice`, `ArraySplice`) for discoverability.

**Comment by sindresorhus** [comment]
Bump :)

**Comment by WeiWen1999** [comment]
Done @sindresorhus

File: test-d/array-at.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+expectType<ArrayAt<TrailingSpreadArray, 2>>({} as string | undefined);
+expectType<ArrayAt<TrailingSpreadArray, 3>>({} as string | undefined);
+expectType<ArrayAt<TrailingSpreadArray, -1>>({} as string | boolean);
+expectType<ArrayAt<TrailingSpreadArray, -2>>({} as boolean | string | number);
+expectType<ArrayAt<TrailingSpreadArray, -3>>({} as boolean | string | number | undefined);
+
+type TrailingSpreadArray2 = [object, number, boolean?, ...string[]];
+expectType<ArrayAt<TrailingSpreadArray2, 0>>({} as object);
+expectType<ArrayAt<TrailingSpreadArray2, 1>>({} as number);
+expectType<ArrayAt<TrailingSpreadArray2, 2>>({} as boolean);
```
Comment (som-sm) [inline test-d/array-at.ts]: This should be `boolean | undefined`, right? Or is it dependent on the `exactOptionalPropertyTypes` compiler option?

**Review by som-sm — COMMENTED** [review]
We’ve recently added several array utility types, like `SplitOnRestElement`, `ExtractRestElement`, `ExcludeRestElement`, and `CollapseRestElement`. Would you mind taking a look at these and seeing if they can help simplify your current approach? I believe there's room for simplification, it'd be really nice if we could not introduce so many new helper types.

I’ll also experiment with this a bit on my side and share my thoughts.

File: source/internal/array.d.ts [hunk 6]
```diff
@@ -60,6 +149,59 @@ export type VariablePartOfArray<T extends UnknownArray> =
...
+		? U
+		: never;
+
+/**
+Returns the trailing static, fixed-length portion of the given middle spread array.
+*/
+export type TrailingStaticPartOfMiddleSpreadArray<T extends UnknownArray> = StaticPartOfLeadingSpreadArray<T>;
+
+/**
+Returns the leading variable, fixed-length portion of the given middle spread array.
```
Comment (sindresorhus) [inline source/internal/array.d.ts]:
```suggestion
Returns the leading static, fixed-length portion of the given middle spread array.
```

File: source/internal/array.d.ts [hunk 7]
```diff
@@ -24,6 +25,91 @@ export type FirstArrayElement<TArray extends UnknownArrayOrTuple> = TArray exten
...
+ //=> [string, number]
+ ```
+ */
+export type RequiredPartOfStaticArray<T extends UnknownArray> =
+	T extends readonly [infer U, ...infer V]
+		? [U, ...RequiredPartOfStaticArray<V>]
+		: [];
+
+/**
+ Returns the optional part of the given array.
```
Comment (sindresorhus) [inline source/internal/array.d.ts]:
```suggestion
Returns the optional part of the given array.
```

same with the rest.

**Comment by sindresorhus** [comment]
A few things:

1. Mixed positive/negative union indices are nt distributing correctly.

```ts
type MixedSign = ArrayAt<[string, number], 0 | -1>;
// current: number | undefined
// expected: string | number
```

This drops the `0` branch and effectively behaves like the negative path only.

2. Optional-leading tuple parts are widened too much for a concrete index when a rest element exists.

```ts
type OptionalCase = ArrayAt<[boolean, number?, {symbolic: true}?, ...string[]], 1>;
// current: number | {symbolic: true} | undefined
// expected: number | undefined
```

Index `1` should not include the later optional element type here. It looks like this is coming from unioning all optional-leading element types instead of just the possible type(s) at the requested index.
